alibi_detect/cd/pytorch/context_aware.py

```diff
@@ -0,0 +1,264 @@
+            ]).mean(0)
+        else:
+            coupling_xx = torch.einsum('ij,ik->ijk', A_0, A_0).mean(0)
+            coupling_yy = torch.einsum('ij,ik->ijk', A_1, A_1).mean(0)
+            coupling_xy = torch.einsum('ij,ik->ijk', A_0, A_1).mean(0)
+        sim_xx = (K[idx_0][:, idx_0]*coupling_xx).sum()
+        sim_yy = (K[idx_1][:, idx_1]*coupling_yy).sum()
+        sim_xy = (K[idx_0][:, idx_1]*coupling_xy).sum()
+        stat = sim_xx + sim_yy - 2*sim_xy
+
+        return stat.cpu(), coupling_xx.cpu(), coupling_yy.cpu(), coupling_xy.cpu()
+
+    def _pick_lam(self, lams: torch.Tensor, K: torch.Tensor, L: torch.Tensor, n_folds: int = 5) -> torch.Tensor:
+        """
+        The conditional mean embedding is estimated as the solution of a regularised regression problem.
+
+        This private method function uses cross validation to select the regularisation parameter that
+        minimises squared error on the out-of-fold instances. The error is a distance in the RKHS and is
+        therefore an MMD-like quantity itself.
+        """
+        n = len(L)
+        fold_size = n // n_folds
+        K, L = K.type(torch.float64), L.type(torch.float64)
+        losses = torch.zeros_like(lams, dtype=torch.float).to(K.device)
+        for fold in range(n_folds):
+            inds_oof = np.arange(n)[(fold*fold_size):((fold+1)*fold_size)]
+            inds_if = np.setdiff1d(np.arange(n), inds_oof)
+            K_if, L_if = K[inds_if][:, inds_if], L[inds_if][:, inds_if]
+            n_if = len(K_if)
+            L_inv_lams = torch.stack(
+                [torch.linalg.inv(L_if + n_if*lam*torch.eye(n_if).to(L.device)) for lam in lams])  # n_lam x n_if x n_if
+            KW = torch.einsum('ij,ljk->lik', K_if, L_inv_lams)
+            lW = torch.einsum('ij,ljk->lik', L[inds_oof][:, inds_if], L_inv_lams)
+            lWKW = torch.einsum('lij,ljk->lik', lW, KW)
+            lWKWl = torch.einsum('lkj,jk->lk', lWKW, L[inds_if][:, inds_oof])  # n_lam x n_oof
+            lWk = torch.einsum('lij,ji->li', lW, K[inds_if][:, inds_oof])  # n_lam x n_oof
+            kxx = torch.ones_like(lWk).to(lWk.device) * torch.max(K)
+            losses += (lWKWl + kxx - 2*lWk).sum(-1)
+        return lams[torch.argmin(losses)]
+
+
+def _sigma_median_diag(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.Tensor:
+    """
+    Private version of the bandwidth estimation function :py:func:`~alibi_detect.utils.pytorch.kernels.sigma_median`,
+    with the +n (and -1) term excluded to account for the diagonal of the kernel matrix.
+
+    Parameters
+    ----------
+    x
+        Tensor of instances with dimension [Nx, features].
+    y
+        Tensor of instances with dimension [Ny, features].
+    dist
+        Tensor with dimensions [Nx, Ny], containing the pairwise distances between `x` and `y`.
+
+    Returns
+    -------
+    The computed bandwidth, `sigma`.
+    """
+    n = min(x.shape[0], y.shape[0])
+    n_median = (np.prod(dist.shape) - n) // 2
```

Comment:
Whilst I doubt it would make a tangible difference, now that I look at this again I think that really it should just be `n_median = np.prod(dist.shape) // 2`